webapp/src/ts/services/cht-datasource.service.ts

```diff
@@ -97,9 +97,21 @@ export class CHTDatasourceService {
     return user?.roles || this.userCtx?.roles;
   }
 
-  async bind <T>(fn: (ctx: DataContext) => T): Promise<T> {
-    await this.isInitialized();
-    return this.dataContext.bind(fn);
+  /**
+   * Binds a cht-datasource function to the data context.
+   * (e.g. `const getPersonWithLineage = this.bind(Person.v1.getWithLineage);`)
+   * @param fn the function to bind. It should accept a data context as the parameter and return another function that
+   * results in a `Promise`.
+   * @returns a "context-aware" version of the function that is bound to the data context and ready to be used
+   */
+  bind<R, F extends (arg?: unknown) => Promise<R>>(fn: (ctx: DataContext) => F):
+    (...p: Parameters<F>) => ReturnType<F> {
+    return (...p) => new Promise((resolve, reject) => this
+      .isInitialized()
+      .then(() => this.dataContext
+        .bind(fn)(...p)
+        .then(resolve)
+        .catch(reject))) as ReturnType<F>;
```

Comment:
Personally I find the autoreturn style with multi line body really difficult to follow. Is this just me?

Secondly I think you can remove an indent from lines 113 and 114. My suggestion would be...

```suggestion
    return (...p) => new Promise((resolve, reject) => {
      return this.isInitialized()
        .then(() => this.dataContext.bind(fn)(...p))
        .then(resolve)
        .catch(reject)) as ReturnType<F>;
    });
```

I _think_ I got all the braces right...

Reply:
:sweat_smile: Generally I don't mind autoreturns for promise chains (and other places where it is pretty clear where things are going).  But, yeah, for this craziness we should make it as readable as possible (even if it means I actually have to type `return` a few more times.... :sob: ).

Going to take this a couple steps further and explicitly return the constructed Promise as well as assigning the contextual function to a variable to avoid the double-parenthesis from the currying. :+1: